<h1 align=center style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
    بسکتبالیست‌ها
</font>
</h1>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        در این سوال قصد داریم تا داده‌های مربوط به عملکرد بازیکنان لیگ بسکتبال NBA را تحلیل کنیم.
        پس کافی است تا کار‌هایی که از شما خواسته شده را مرحله به مرحله انجام دهید.
    </font>
</p>

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        مجموعه‌داده
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: right;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
        مجموعه‌داده‌ای که در احتیار شما قرار گرفته است شامل اطلاعات عملکرد بازیکنان NBA در فصل‌های ۲۰۰۰ الی ۲۰۰۹ است.
    این مجموعه شامل ۱۸۳۰ سطر و ۱۸ ستون می‌باشد که هر سطر آن شامل اطلاعات یک بازیکن در یک فصل است. توضیحات مربوط به ستون‌ها نیز در جدول زیر آمده است.
    </font>
</p>

<center>
<div dir=rtl style="direction: rtl;line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|ستون|توضیحات|
|:------:|:---:|
|<code>PlAYER</code>|نام بازیکن|
|<code>TEAM</code>|مخفف اسم تیم بازیکن در آن فصل|
|<code>YEAR</code>|سال (فصل مسابقات)|
|<code>GP</code>|تعداد بازی‌های انجام شده|
|<code>MIN</code>|میانگین دقایق بازی کرده|
|<code>PTS</code>|میانگین امتیاز|
|<code>FGM</code>|میانگین تعداد پرتاب‌های ۲ امتیازی موفق|
|<code>FGA</code>|میانگین تعداد کل پرتاب‌های ۲ امتیازی|
|<code>3PM</code>|میانگین تعداد پرتاب‌های ۳ امتیازی موفق|
|<code>3PA</code>|میانگین تعداد کل پرتاب‌های ۳ امتیازی|
|<code>FTM</code>|میانگین تعداد پرتاب‌های آزاد (پنالتی) موفق|
|<code>FTA</code>|میانگین تعداد کل پرتاب‌های آزاد (پنالتی)|
|<code>OREB</code>|میانگین تعداد ریباند‌ها در حمله|
|<code>DREB</code>|میانگین تعداد ریباند‌ها در دفاع|
|<code>AST</code>|میانگین تعداد پاس‌های منجر به امتیاز|
|<code>STL</code>|میانگین تعداد دفعات توپ‌ربایی|
|<code>BLK</code>|میانگین تعداد بلاک‌ها|
|<code>TOV</code>|میانگین تعداد دفعات لو دادن توپ|
    
</font>
</div>
</center>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
   با استفاده از سلول زیر کتابخانه‌های مورد نیاز خود را <code>import</code> کنید و سپس مجموعه‌داده را که در فایلی به نام <code>NBA2000-2009.csv</code> قرار دارد، بخوانید. 
</font>
</p>

In [5]:
import pandas as pd

df = pd.read_csv("/home/reza/Downloads/NBA/Data/NBA2000-2009.csv")
df.head()

PLAYER TEAM  YEAR  GP   MIN   PTS   FGM   FGA  3PM  3PA  FTM  \
0     Allen Iverson  PHI  2000  71  42.0  31.1  10.7  25.5  1.4  4.3  8.2   
1  Jerry Stackhouse  DET  2000  80  40.2  29.8   9.7  24.1  2.1  5.9  8.3   
2  Shaquille O'Neal  LAL  2000  74  39.5  28.7  11.0  19.2  0.0  0.0  6.7   
3       Kobe Bryant  LAL  2000  68  40.9  28.5  10.3  22.2  0.9  2.9  7.0   
4      Vince Carter  TOR  2000  75  39.7  27.6  10.2  22.1  2.2  5.3  5.1   

    FTA  OREB  DREB  AST  STL  BLK  TOV  
0  10.1   0.7   3.1  4.6  2.5  0.3  3.3  
1  10.1   1.2   2.7  5.1  1.2  0.7  4.1  
2  13.1   3.9   8.8  3.7  0.6  2.8  2.9  
3   8.2   1.5   4.3  5.0  1.7  0.6  3.2  
4   6.7   2.3   3.2  3.9  1.5  1.1  2.2

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت اول
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         لیستی از بازیکن‌ها و فصل‌ها را بدست آورید که در آن فصل، آن بازیکن در حداقل ۲ تا از ۵ ملاک اصلی آمار دو رقمی (بزرگتر مساوی ۱۰) ثبت کرده باشد. 
        <br>
        منظور از ۵ ملاک اصلی میانگین امتیاز، میانگین تعداد ریباند (مجموع ریباند در دفاع و حمله)، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها است.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>double_double</code> قرار دهید.
    این دیتافریم باید شامل ۷ ستون باشد که ستون‌ها از چپ به راست به ترتیب نام بازیکن، سال، میانگین امتیاز، میانگین تعداد پاس‌های منجر به گل، میانگین تعداد ریباند‌ها، میانگین تعداد بلاک‌ها و میانگین تعداد توپ‌ربایی‌ها باشند. همچنین کل دیتافریم باید به ترتیب برحسب سال و نام بازیکن به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|YEAR|PTS|AST|REB|STL|BLK|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|Antonio Davis|2000|13.7|1.4|10.1|0.3|1.9|
|Antonio McDyess|2000|20.8|2.1|12.0|0.6|1.5|
    
</font>
</div>
</center>

In [19]:
df['REB'] = df['OREB'] + df['DREB']

def count_double_digit_stats(row):
    stats = [
        row['PTS'] >= 10,
        row['AST'] >= 10,
        row['REB'] >= 10,
        row['BLK'] >= 10,
        row['STL'] >= 10
    ]
    return sum(stats) >= 2

double_double = df[df.apply(count_double_digit_stats, axis=1)]

double_double = double_double[['PLAYER', 'YEAR', 'PTS', 'AST', 'REB', 'BLK', 'STL']]

double_double = double_double.sort_values(by=['YEAR', 'PLAYER'])



double_double



PLAYER  YEAR   PTS   AST   REB  BLK  STL
55      Antonio Davis  2000  13.7   1.4  10.1  1.9  0.3
19    Antonio McDyess  2000  20.8   2.1  12.0  1.5  0.6
5        Chris Webber  2000  27.1   4.2  11.1  1.7  1.3
92    Dikembe Mutombo  2000  10.0   1.0  13.5  2.7  0.4
23        Elton Brand  2000  20.1   3.2  10.1  1.6  1.0
...               ...   ...   ...   ...   ...  ...  ...
1670   Gerald Wallace  2009  18.2   2.1  10.1  1.1  1.5
1686       Steve Nash  2009  16.5  11.0   3.3  0.1  0.5
1672       Tim Duncan  2009  17.9   3.2  10.1  1.5  0.6
1701      Troy Murphy  2009  14.6   2.1  10.2  0.5  1.0
1655    Zach Randolph  2009  20.8   1.8  11.8  0.4  1.0

[88 rows x 7 columns]

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت دوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         باارزش‌ترین بازیکنان این ۱۰ فصل را مشخص کنید.
        ارزش یک بازیکن در یک فصل با استفاده از فرمول زیر بدست می‌آید:
    </font>
    $$ (PTS + OREB + DREB + AST + STL + BLK) - (TOV + Missed FG + Missed FT) $$
    منظور از <code>Missed FG</code> و <code>Missed FT</code> به ترتیب میانگین تعداد پرتاب‌های ۲ امتیازی و آزاد از دست رفته است.  
    <br>
    ارزش یک بازیکن در کل ۱۰ فصل برابر است با میانگین ارزش او در هر فصل.
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>best_player</code> قرار دهید.
    این دیتافریم شامل دو ستون است که یکی نام بازیکن و دومی ارزش آن بازیکن است که تا دو رقم اعشار باید گرد شود. همچنین این دیتافریم باید بر اساس ارزش به صورت نزولی و سپس بر اساس نام بازیکن به‌صورت صعودی مرتب باشد.
    دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|PLAYER|VALUE|
|:------:|:---:|
|Kevin Garnett|29.74|
|LeBron James|27.9|
    
</font>
</div>
</center>

In [18]:
df['Missed_FG'] = df['FGA'] - df['FGM']
df['Missed_FT'] = df['FTA'] - df['FTM']

df['Value'] = (df['PTS'] + df['OREB'] + df['DREB'] + df['AST'] + df['STL'] + df['BLK']) - (df['TOV'] + df['Missed_FG'] + df['Missed_FT'])

best_player = df.groupby('PLAYER')['Value'].mean().reset_index()

best_player['Value'] = best_player['Value'].round(2)

best_player = best_player.sort_values(by=['Value', 'PLAYER'], ascending=[False, True])

best_player



PLAYER  Value
316         Kevin Garnett  29.74
335          LeBron James  27.90
98             Chris Paul  26.57
495      Shaquille O'Neal  26.55
169           Dwyane Wade  26.12
..                    ...    ...
432          Quincy Douby   3.60
512            Steve Kerr   3.60
65       Brian Scalabrine   3.30
415  Nikoloz Tskitishvili   3.30
478          Sam Mitchell   3.15

[571 rows x 2 columns]

<h2 align=right style="line-height:200%;font-family:vazir;color:#0099cc">
    <font face="vazir" color="#0099cc">
        قسمت سوم
    </font>
</h2>

<p dir=rtl style="direction: rtl;text-align: justify;line-height:200%;font-family:vazir;font-size:medium">
    <font face="vazir" size=3>
         در هر سال چه تیمی بیشترین میانگین امتیاز را ثبت کرده است؟
        <br>
        توجه داشته باشید که میانگین امتیاز یک تیم برابر است با مجموع میانگین امتیاز بازیکن‌هایش.
    </font>
    <br>
    <font face="vazir" color="red">
        خروجی:
    </font>
    نتیجه‌ی درخواست را در دیتافریمی با نام <code>max_PTS_of_year</code> قرار دهید.
    این دیتافریم باید شامل سه ستون باشد که ستون‌ها از چپ به راست به ترتیب سال، نام تیم و میانگین امتیاز است که امتیاز‌ها باید تا دو رقم اعشار گرد شوند. همچنین کل دیتافریم باید برحسب سال به صورت صعودی مرتب باشد. دو سطر اول این دیتافریم به شکل زیر است:
</p>

<center>
<div style="line-height:200%;font-family:vazir;font-size:medium">
<font face="vazir" size=3>
    
|YEAR|TEAM|PTS|
|:------:|:---:|:---:|
|2000|SAC|100.2|
|2001|LAL|95.9|
    
</font>
</div>
</center>

In [17]:
team_pts_per_year = df.groupby(['YEAR', 'TEAM'])['PTS'].sum().reset_index()

team_pts_per_year['PTS'] = team_pts_per_year['PTS'].round(2)

max_PTS_of_year = team_pts_per_year.loc[team_pts_per_year.groupby('YEAR')['PTS'].idxmax()]

max_PTS_of_year = max_PTS_of_year.sort_values(by='YEAR')

max_PTS_of_year

YEAR TEAM    PTS
22   2000  SAC  100.2
41   2001  LAL   95.9
77   2002  ORL   97.0
92   2003  DEN   96.0
117  2004  BOS  106.2
175  2005  WAS   96.8
198  2006  PHX  105.4
212  2007  DEN  110.7
242  2008  DEN   97.3
286  2009  OKC  102.5

<h2 dir=rtl align=right style="line-height:200%;font-family:vazir;color:#0099cc">
<font face="vazir" color="#0099cc">
<b>سلول جواب‌ساز</b>
</font>
</h2>

<p dir=rtl style="direction: rtl; text-align: justify; line-height:200%; font-family:vazir; font-size:medium">
<font face="vazir" size=3>
    برای ساخته‌شدن فایل <code>result.zip</code> سلول زیر را اجرا کنید. توجه داشته باشید که پیش از اجرای سلول زیر تغییرات اعمال شده در نت‌بوک را ذخیره کرده باشید (<code>ctrl+s</code>) در غیر این صورت، در پایان مسابقه نمره شما به صفر تغییر خواهد کرد.
    <br>
    همچنین اگر از کولب برای اجرای این فایل نوت‌بوک استفاده می‌کنید، قبل از ارسال فایل <code>result.zip</code>، آخرین نسخه‌ی نوت‌بوک خود را دانلود کرده و داخل فایل ارسالی قرار دهید.
</font>

In [20]:
import zipfile
import os

if not os.path.exists(os.path.join(os.getcwd(), 'NBA.ipynb')):
    %notebook -e NBA.ipynb

double_double.to_csv('double_double.csv', index=False)
best_player.to_csv('best_player.csv', index=False)
max_PTS_of_year.to_csv('max_PTS_of_year.csv', index=False)

def compress(file_names):
    print("File Paths:")
    print(file_names)
    compression = zipfile.ZIP_DEFLATED
    with zipfile.ZipFile("result.zip", mode="w") as zf:
        for file_name in file_names:
            zf.write('./' + file_name, file_name, compress_type=compression)

file_names = ['double_double.csv', 'best_player.csv', 'max_PTS_of_year.csv', 'NBA.ipynb']
compress(file_names)

NameError: name 'os' is not defined